### PR TITLE
Add EventTarget & EventBus

### PR DESCRIPTION
**Pull request by PatriceJiang**
Impl EventBus & EventTarget

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/event/EventBus.h [hunk 1]
```diff
@@ -0,0 +1,1165 @@
...
+    class EventType final : public ebus::TargetEventBaseHandler<EmitterType, static_cast<int>(EventValue), decltype(EventValue)> { \
+    public:                                                                                                                        \
+        using base_type = ebus::TargetEventBaseHandler<EmitterType, static_cast<int>(EventValue), decltype(EventValue)>;           \
+        using event_type = ebus::Event<EventType>;                                                                                 \
+        using persist_function_type = std::function<void(EmitterType *, event_type *)>;                                            \
+        using handler_type = std::function<void(event_type *)>;                                                                    \
+        constexpr static const char *EVENT_NAME = #EventType;                                                                      \
+        const char *getEventType() const { return typeid(typename base_type::event_enum_type).name(); }                            \
```
Comment (dumganhar) [inline native/cocos/event/EventBus.h]: Could `getEventType` be static?

File: native/cocos/event/EventBus.h [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+        using base_type = ebus::TargetEventBaseHandler<EmitterType, static_cast<int>(EventValue), decltype(EventValue), __VA_ARGS__>;           \
+        using event_type = ebus::Event<EventType>;                                                                                              \
+        using persist_function_type = std::function<void(EmitterType *, event_type *)>;                                                         \
+        using handler_type = std::function<void(event_type *)>;                                                                                 \
+        constexpr static const char *EVENT_NAME = #EventType;                                                                                   \
+        const char *getEventType() const { return typeid(typename base_type::event_enum_type).name(); }                                         \
+    }
+
+#define DECLARE_TARGET_EVENT1(EventType, EmitterType, EventValue, arg_type_0) \
+    DECLARE_TARGET_EVENT(EventType, EmitterType, EventValue, arg_type_0)
```
Comment (dumganhar) [inline native/cocos/event/EventBus.h]: Could we just use DECLARE_TARGET_EVENT for arguments and DECLARE_TARGET_EVENT0 without arguments?

File: native/cocos/bindings/event/EventDispatcher.cpp [hunk 3]
```diff
@@ -326,38 +358,24 @@ void EventDispatcher::dispatchTickEvent(float /*dt*/) {
...
+void EventDispatcher::dispatchResizeEvent(int width, int height, uint32_t windowId) {
     se::AutoHandleScope scope;
     if (!jsResizeEventObj) {
         jsResizeEventObj = se::Object::createPlainObject();
         jsResizeEventObj->root();
     }
 
+    if (windowId != UINT32_MAX) {
+        jsResizeEventObj->setProperty("width", se::Value(width));
+    }
```
Comment (dumganhar) [inline native/cocos/bindings/event/EventDispatcher.cpp]: What does this logic mean? width ? windowId ?

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

File: native/cocos/engine/Engine.cpp [hunk 4]
```diff
@@ -309,89 +293,45 @@ void Engine::tick() {
...
-    return isHandled;
-}
-
-bool Engine::dispatchWindowEvent(const WindowEvent &ev) {
+bool Engine::redirectWindowEvent(const WindowEvent &ev) {
     bool isHandled = false;
     if (ev.type == WindowEvent::Type::SHOW ||
         ev.type == WindowEvent::Type::RESTORED) {
-        emit(static_cast<int>(ON_RESUME));
+        static_cast<BaseEngine*>(this)->emit<EngineStatusChange>(ON_RESUME);
```
Comment (qiuguohua) [inline native/cocos/engine/Engine.cpp]: Why need to convert to a base class when engine inherits from BaseEngine and should be able to call it directly?

File: native/cocos/platform/SDLHelper.cpp [hunk 5]
```diff
@@ -160,43 +160,43 @@ int SDLHelper::init() {
...
-void SDLHelper::dispatchWindowEvent(IEventDispatch *delegate, uint32_t windowId, const SDL_WindowEvent &wevent) {
+void SDLHelper::dispatchWindowEvent(uint32_t windowId, const SDL_WindowEvent &wevent) {
     WindowEvent ev;
     ev.windowId = windowId;
 
     switch (wevent.event) {
         case SDL_WINDOWEVENT_SHOWN: {
             ev.type = WindowEvent::Type::SHOW;
-            delegate->dispatchEvent(ev);
+            cc::event::broadcast<cc::events::WindowEvent>(ev);
```
Comment (qiuguohua) [inline native/cocos/platform/SDLHelper.cpp]: It is not recommended to send messages directly here. The purpose of this design is to encapsulate the event distribution in a higher level. If you change it, you can just change the base class directly.
According to the current modification, it is necessary to change all platforms, which is not conducive to cross-platform

File: native/cocos/engine/Engine.cpp (hunk 4 above)
Comment (PatriceJiang) [reply native/cocos/engine/Engine.cpp]: It is a template function that is not inherited.

File: native/cocos/application/CocosApplication.cpp [hunk 6]
```diff
@@ -63,22 +59,28 @@ int CocosApplication::init() {
...
+                this->onResume();
+                break;
+            case BaseEngine::ON_PAUSE:
+                this->onPause();
+                break;
+            case BaseEngine::ON_CLOSE:
+                this->onClose();
+                break;
+            default:
+                assert(false);
```
Comment (minggo) [inline native/cocos/application/CocosApplication.cpp]: Please use CC_ASSERT

File: native/cocos/bindings/event/CustomEventTypes.h [hunk 7]
```diff
@@ -25,17 +25,3 @@
 
 #pragma once
 
-#define EVENT_COME_TO_FOREGROUND "event_come_to_foreground"
```
Comment (minggo) [inline native/cocos/bindings/event/CustomEventTypes.h:28]: Can just remove this file?

File: native/cocos/core/Root.h [hunk 8]
```diff
@@ -324,10 +328,10 @@ class Root final {
     uint32_t _fixedFPS{0};
     bool _useDeferredPipeline{false};
     bool _usesCustomPipeline{true};
-    CallbacksInvoker *_eventProcessor{nullptr};
     IXRInterface *_xr{nullptr};
-    uint32_t _windowDestroyEventId{0};
-    uint32_t _windowResumeEventId{0};
+    mutable cc::event::Listener<events::WindowDestroy> _windowDestroyEventId;
+    mutable cc::event::Listener<events::WindowRecreated> _windowRecreatedEventId;
```
Comment (minggo) [inline native/cocos/core/Root.h]: Should rename to listener.

File: native/cocos/core/assets/Asset.h [hunk 9]
```diff
@@ -32,13 +32,13 @@
 #include "base/Macros.h"
 #include "core/Types.h"
 #include "core/data/Object.h"
-#include "core/event/CallbacksInvoker.h"
+#include "event/Event.h"
```
Comment (minggo) [inline native/cocos/core/assets/Asset.h]: Please include from `core`.

**Review by qiuguohua — APPROVED** [review]

File: native/cocos/core/assets/Asset.h [hunk 10]
```diff
@@ -32,13 +32,13 @@
...
 
 namespace cc {
 
 class Node;
 
-class Asset : public CCObject, public EventTarget {
+class Asset : public CCObject , public cc::event::EventTarget {
```
Comment (minggo) [inline native/cocos/core/assets/Asset.h]: `cc::` is not needed.

File: native/cocos/core/assets/EffectAsset.h [hunk 11]
```diff
@@ -581,6 +582,8 @@ class EffectAsset final : public Asset {
     void initDefault(const ccstd::optional<ccstd::string> &uuid) override;
     bool validate() const override;
 
+private:
+    BaseEngine::EngineStatusChange::EventID _engineListener;
```
Comment (minggo) [inline native/cocos/core/assets/EffectAsset.h]: Please rename to _engineEventID. And it is better put `private` after `protected`.

File: native/cocos/core/assets/Material.h [hunk 12]
```diff
@@ -87,6 +88,8 @@ struct IMaterialInfo {
 };
 
 class Material : public Asset {
+    IMPL_EVENT_TARGET(Material)
+    DECLARE_TARGET_EVENT0(Material_Passes_Updated, Material)
```
Comment (minggo) [inline native/cocos/core/assets/Material.h]: Please do not use `_` to connect words.

File: native/cocos/core/event/EventTypesToJS.cpp [hunk 13]
```diff
@@ -31,39 +31,37 @@ namespace cc {
...
-const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_LOCAL_SH_DESCRIPTORS{113};    //{"MODEL_UPDATE_LOCAL_SH_DESCRIPTORS"};
-const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_WORLD_BOUND_DESCRIPTORS{114}; //{"MODEL_UPDATE_WORLD_BOUND_DESCRIPTORS"};
-const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_INSTANCED_ATTRIBUTES{115};    //{"MODEL_UPDATE_INSTANCED_ATTRIBUTES"};
-const CallbacksInvoker::KeyType EventTypesToJS::MODEL_GET_MACRO_PATCHES{116};              //{"MODEL_GET_MACRO_PATCHES"};
+// const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_TRANSFORM{110};            //{"MODEL_UPDATE_TRANSFORM"};
+// const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_UBO{111};                  //{"MODEL_UPDATE_UBO"};
+// const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_LOCAL_DESCRIPTORS{112};    //{"MODEL_UPDATE_LOCAL_DESCRIPTORS"};
+// const CallbacksInvoker::KeyType EventTypesToJS::MODEL_UPDATE_INSTANCED_ATTRIBUTES{113}; //{"MODEL_UPDATE_INSTANCED_ATTRIBUTES"};
+// const CallbacksInvoker::KeyType EventTypesToJS::MODEL_GET_MACRO_PATCHES{114};           //{"MODEL_GET_MACRO_PATCHES"};
```
Comment (minggo) [inline native/cocos/core/event/EventTypesToJS.cpp]: Please remove unneeded codes.

File: native/cocos/core/event/EventTypesToJS.h [hunk 14]
```diff
@@ -23,44 +23,4 @@
  THE SOFTWARE.
  ****************************************************************************/
 
-#pragma once
```
Comment (minggo) [inline native/cocos/core/event/EventTypesToJS.h:26]: Can remove this file?

File: native/cocos/event/Event.h [hunk 15]
```diff
@@ -1,48 +1,53 @@
...
-     */
-    virtual void dispatchEvent(const OSEvent& ev) = 0;
-    /**
-     * Dispatch touch event interface.
-     */
-    virtual void dispatchTouchEvent(const TouchEvent& ev) = 0;
-};
-
-} // namespace cc
+#ifdef SWIGCOCOS
```
Comment (minggo) [inline native/cocos/event/Event.h]: Event system is better to move to core.

File: native/cocos/application/CocosApplication.cpp (hunk 6 above)
Comment (PatriceJiang) [reply native/cocos/application/CocosApplication.cpp]: fixed

File: native/cocos/core/Root.h (hunk 8 above)
Comment (PatriceJiang) [reply native/cocos/core/Root.h]: fixed

File: native/cocos/bindings/event/CustomEventTypes.h (hunk 7 above)
Comment (PatriceJiang) [reply native/cocos/bindings/event/CustomEventTypes.h:28]: fixed

File: native/cocos/core/assets/Asset.h (hunk 10 above)
Comment (PatriceJiang) [reply native/cocos/core/assets/Asset.h]: fixed

File: native/cocos/core/assets/Material.h (hunk 12 above)
Comment (PatriceJiang) [reply native/cocos/core/assets/Material.h]: fixed

File: native/cocos/core/event/EventTypesToJS.cpp (hunk 13 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventTypesToJS.cpp]: fixed

File: native/cocos/core/event/EventTypesToJS.h (hunk 14 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventTypesToJS.h:26]: fixed

File: native/cocos/engine/EngineEvents.h [hunk 16]
```diff
@@ -0,0 +1,292 @@
+#pragma once
```
Comment (minggo) [inline native/cocos/engine/EngineEvents.h:25]: Please add license.

File: native/cocos/core/assets/EffectAsset.h (hunk 11 above)
Comment (PatriceJiang) [reply native/cocos/core/assets/EffectAsset.h]: fxied

File: native/cocos/event/EventBus.cpp [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+#include "EventBus.h"
+#include <algorithm>
+#include <cassert>
+#include <unordered_map>
+#include <vector>
```
Comment (minggo) [inline native/cocos/event/EventBus.cpp]: Please use ccstd::vector/ccstd::unordered_map instead. And it seems they are not used?

File: native/cocos/event/EventBus.h [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+#pragma once
+
+#include <cassert>
```
Comment (minggo) [inline native/cocos/event/EventBus.h]: Just to use `assert`? If so, please use `CC_ASSERT` instead. And it is included in .cpp too.

File: native/tools/simulator/libsimulator/lib/CustomAppEvent.cpp [hunk 19]
```diff
@@ -30,30 +30,3 @@
 //
 
 #include "CustomAppEvent.h"
-
-CustomAppEvent::CustomAppEvent(const std::string& eventName, int type)
```
Comment (minggo) [inline native/tools/simulator/libsimulator/lib/CustomAppEvent.cpp:34]: Can remove it?

**Review by minggo — APPROVED** [review]

File: native/tools/simulator/libsimulator/lib/CustomAppEvent.cpp (hunk 19 above)
Comment (PatriceJiang) [reply native/tools/simulator/libsimulator/lib/CustomAppEvent.cpp:34]: removed

File: native/cocos/event/EventBus.h (hunk 18 above)
Comment (PatriceJiang) [reply native/cocos/event/EventBus.h]: replaced

File: native/cocos/event/EventBus.cpp (hunk 17 above)
Comment (PatriceJiang) [reply native/cocos/event/EventBus.cpp]: Removed

File: native/cocos/engine/EngineEvents.h (hunk 16 above)
Comment (PatriceJiang) [reply native/cocos/engine/EngineEvents.h:25]: Added

File: native/cocos/event/Event.h (hunk 15 above)
Comment (PatriceJiang) [reply native/cocos/event/Event.h]: Done

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

File: native/cocos/bindings/event/EventDispatcher.cpp [hunk 20]
```diff
@@ -218,7 +241,7 @@ void EventDispatcher::dispatchMouseEvent(const MouseEvent &mouseEvent) {
     se::ValueArray args;
     args.emplace_back(se::Value(jsMouseEventObj));
     EventDispatcher::doDispatchJsEvent(jsFunctionName, args);
-    EventDispatcher::dispatchCustomEvent(eventName, 0);
+    // EventDispatcher::dispatchCustomEvent(eventName, 0);
```
Comment (dumganhar) [inline native/cocos/bindings/event/EventDispatcher.cpp]: dispatchCustomEvent isn't needed now?

Comment (dumganhar) [reply native/cocos/bindings/event/EventDispatcher.cpp]: If it's not needed now, just delete this line.

File: native/cocos/engine/Engine.cpp [hunk 21]
```diff
@@ -129,14 +133,10 @@ int32_t Engine::init() {
...
     BasePlatform *platform = BasePlatform::getPlatform();
-    platform->setHandleEventCallback(
-        std::bind(&Engine::handleEvent, this, std::placeholders::_1)); // NOLINT(modernize-avoid-bind)
-
-    platform->setHandleTouchEventCallback(
-        std::bind(&Engine::handleTouchEvent, this, std::placeholders::_1)); // NOLINT(modernize-avoid-bind)
 
     se::ScriptEngine::getInstance()->addRegisterCallback(setCanvasCallback);
-    emit(static_cast<int>(ON_START));
+    // emit(static_cast<int>(ON_START));
```
Comment (dumganhar) [inline native/cocos/engine/Engine.cpp]: Remove unused code.

File: native/cocos/engine/Engine.h [hunk 22]
```diff
@@ -158,7 +134,8 @@ class CC_DLL Engine : public BaseEngine {
     BuiltinResMgr *_builtinResMgr{nullptr};
     ProgramLib *_programLib{nullptr};
 
-    std::map<OSEventType, EventCb> _eventCallbacks;
+    cc::event::Listener<events::WindowEvent> _windowEvents;
```
Comment (dumganhar) [inline native/cocos/engine/Engine.h]: _windowEvents -> _windowEventListener

Comment (PatriceJiang) [reply native/cocos/engine/Engine.h]: renamed

File: native/cocos/engine/Engine.cpp (hunk 21 above)
Comment (PatriceJiang) [reply native/cocos/engine/Engine.cpp]: removed

File: native/cocos/bindings/event/EventDispatcher.cpp (hunk 20 above)
Comment (PatriceJiang) [reply native/cocos/bindings/event/EventDispatcher.cpp]: deleted

**Review by dumganhar — APPROVED** [review]

**Comment by PatriceJiang** [comment]
@cocos-robot run test cases

**Pull request merged**